pkg/cache/references.go

```diff
@@ -0,0 +1,130 @@
+package cache
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/argoproj/gitops-engine/pkg/utils/kube"
+)
+
+var (
+	pvcNameRegex = regexp.MustCompile(`.+-(.+)-\d+`)
+)
+
+func (c *clusterCache) resolveResourceReferences(un *unstructured.Unstructured) ([]metav1.OwnerReference, func(kube.ResourceKey) bool) {
+	var isChildRef func(_ kube.ResourceKey) bool
+	ownerRefs := un.GetOwnerReferences()
+	gvk := un.GroupVersionKind()
+	// Special case for endpoint. Remove after https://github.com/kubernetes/kubernetes/issues/28483 is fixed
+	if gvk.Group == "" && gvk.Kind == kube.EndpointsKind && len(un.GetOwnerReferences()) == 0 {
```

Comment:
nitpick: all these `if`s can probably be turned into a `switch` for better readability.

Reply:
Thank you for the suggestion! Switch is actually easier to read. Updated.